File: algorithms/0134.gas_station/gas_station.cpp
class Solution {
public:
    int canCompleteCircuit(const vector<int>& gas, const vector<int>& cost) {
        const size_t len = gas.size();
        int cur_sum = 0, sum = 0;
        size_t it = 0;
        for (size_t i = 0; i < len; ++i) {
            const int diff = gas[i] - cost[i];
            cur_sum += diff;
            sum += diff;
            if (cur_sum < 0) {
                cur_sum = 0;
                it = i + 1;
            }
        }
        if (sum < 0) {
            return -1;
        }
        return it;
    }
};

File: algorithms/0134.gas_station/gas_station.rs
impl Solution {
pub fn can_complete_circuit(gas: Vec<i32>, cost: Vec<i32>) -> i32 {
    let mut sum = 0;
    let mut curr_sum = 0;
    let mut start = 0;
    for i in 0..gas.len() {
        let diff = gas[i] - cost[i];
        sum += diff;
        curr_sum += diff;
        if curr_sum < 0 {
            start = i as i32 + 1;
            curr_sum = 0;
        }
    }
    if sum < 0 {
        return -1;
    }
    start
}
}
